File: Source/AlphaTab/Rendering/Effects/LetRingEffectInfo.cs
using AlphaTab.Model;
using AlphaTab.Rendering.Glyphs;

namespace AlphaTab.Rendering.Effects
{
    internal class LetRingEffectInfo : IEffectBarRendererInfo
    {
        public string EffectId => "let-ring";
        public bool CanShareBand => false;
        public virtual bool HideOnMultiTrack => false;

        public bool ShouldCreateGlyph(Settings settings, Beat beat)
        {
            return beat.IsLetRing;
        }

        public EffectBarGlyphSizing SizingMode => EffectBarGlyphSizing.GroupedOnBeat;

        public EffectGlyph CreateNewGlyph(BarRendererBase renderer, Beat beat)
        {
            return new LineRangedGlyph("LetRing");
        }

        public virtual bool CanExpand(Beat from, Beat to)
        {
            return true;
        }
    }
}

File: packages/alphatab/src/rendering/effects/LetRingEffectInfo.ts
import type { Beat } from '@src/model/Beat';
import type { BarRendererBase } from '@src/rendering/BarRendererBase';
import { EffectBarGlyphSizing } from '@src/rendering/EffectBarGlyphSizing';
import type { EffectGlyph } from '@src/rendering/glyphs/EffectGlyph';
import { LineRangedGlyph } from '@src/rendering/glyphs/LineRangedGlyph';
import { EffectBarRendererInfo } from '@src/rendering/EffectBarRendererInfo';
import type { Settings } from '@src/Settings';
import { NotationElement } from '@src/NotationSettings';

/**
 * @internal
 */
export class LetRingEffectInfo extends EffectBarRendererInfo {
    public get notationElement(): NotationElement {
        return NotationElement.EffectLetRing;
    }

    public get canShareBand(): boolean {
        return false;
    }

    public get hideOnMultiTrack(): boolean {
        return false;
    }

    public shouldCreateGlyph(_settings: Settings, beat: Beat): boolean {
        return beat.isLetRing;
    }

    public get sizingMode(): EffectBarGlyphSizing {
        return EffectBarGlyphSizing.GroupedOnBeat;
    }

    public createNewGlyph(_renderer: BarRendererBase, _beat: Beat): EffectGlyph {
        return new LineRangedGlyph('LetRing');
    }

    public canExpand(_from: Beat, _to: Beat): boolean {
        return true;
    }
}
